pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java

```diff
@@ -123,7 +123,12 @@ enum MetadataKey {
     NUM_SEGMENTS_PRUNED_BY_LIMIT("numSegmentsPrunedByLimit", MetadataValueType.INT),
     NUM_SEGMENTS_PRUNED_BY_VALUE("numSegmentsPrunedByValue", MetadataValueType.INT),
     EXPLAIN_PLAN_NUM_EMPTY_FILTER_SEGMENTS("explainPlanNumEmptyFilterSegments", MetadataValueType.INT),
-    EXPLAIN_PLAN_NUM_MATCH_ALL_FILTER_SEGMENTS("explainPlanNumMatchAllFilterSegments", MetadataValueType.INT);
+    EXPLAIN_PLAN_NUM_MATCH_ALL_FILTER_SEGMENTS("explainPlanNumMatchAllFilterSegments", MetadataValueType.INT),
+    // TODO: Add a mechanism to pass generic metadata without adding them to the DataTable's metadata
+    // This metadata field is only required for emitting a metric to identify whether queries using MV columns (as
+    // identifiers or via transform) are present in the Selection Order-By queries. This was the only place where
+    // such a metadata could be added in the existing interfaces even though we don't need this on the Broker side.
+    QUERY_HAS_MV_SELECTION_ORDER_BY("queryHasMVSelectionOrderBy", MetadataValueType.STRING);
```

Comment:
I'd suggest not adding this because this is a temporary step to catch the bad queries, and we need to remove this later. If we add it as a `DataTable.MetadataKey`, we cannot remove it in the future. See the javadoc for this enum for why it cannot be removed

Reply:
I agree. The current code is structured in a way that this is the only mechanism to communicate back such info from operators to QueryScheduler code in server which emits / logs metrics.

This is a problem as it forces (like in this case) to add a metadata key when we don't even need one.

Two potential solutions but important to do them cleanly/ right abstraction

- either change APIs in way to propagate down ServerMetrics into operators (processQuery API  ....)
- have a map in DataTable which is not part of serialization but for transient purpose like above

Thoughts @somandal  @Jackie-Jiang

Reply:
Good point! Didn't notice `ServerMetrics` is not available in query path.
I'd suggest passing down `ServerMetrics` because it can be useful for other features when we want to track certain stats for query runtime. `ServerMetrics` is available in `ServerQueryExecutorV1Impl`, and we need to pass it down to the plan maker all the way to the operators

Reply:
hey @Jackie-Jiang, @siddharthteotia and I discussed the `ServerMetrics` change and do think it is a good idea in general. One problem is that we need an additional way to correlate a query with this metric, and for that purpose we actually added a log in `QueryScheduler` in this existing PR to capture the value of this flag. This is especially important for high QPS use cases where we may need to comb through hundreds of logs on the Broker side to find the possible offending queries. In addition to that, it may not be desirable to emit a metric for every segment (since the operators are at a segment level) and instead just emit once per query.

Thinking some more we have a couple of ideas to address this (have both a log and a metric and avoid doing this for every segment belonging to the query):

Option 1 (preferred method):

- Modify the `QueryContext` to also contain the request ID from the `ServerQueryRequest`. This will then be available throughout all the plan nodes and operators. We can then use this request ID for logging purposes.
- To avoid emitting a metric and log in the `SelectionOrderByOperator` but instead do this at the `InstanceResponseOperator` level where we will have the combined `IntermediateResultsBlock` available. This will allow us to emit the metric and log only once per query per server. The `SelectionOrderByOperator` will still detect this scenario and set it in the `IntermediateResultsBlock` as is being done in the current PR.
- For now we can avoid modifying all the operators and plan nodes to take the `ServerMetrics` as input and instead only pass it down up to the `InstanceResponsePlanNode` and `InstanceResponseOperator`. In the future it should be trivial to extend all operators and plan nodes to have the `ServerMetrics` if the need arises.

Option 2

- Add a transient metadata object to the `IntermediateResultsBlock` and potentially `DataTable` where we can store all the transient metadata fields which aren't serializable and will be ignored when creating the serialized `DataTable`. The drawback with this approach is that certain operators like Filter don't use `IntermediateResultsBlock` directly and we'll need to add getters/setters to all the intermediate block types and copy these over. We still only want to emit the metric and log just once per query.


Let us know your thoughts @Jackie-Jiang

Reply:
May be there is no need to log the requestID etc to be able to correlate with the offending query on broker. All we can do is simply log the ORDER BY expressions and simply grep for that in the broker log and we can find the exact query.

`"Table <T> has MV column in ORDER BY. Expressions: <>"`

So, we just need the following change

- Pass ServerMetrics to plan maker
- From PlanMaker pass to SelectionOrderByPlanNode / InstanceResponsePlanNode
- Emit metric
- Log MV ORDER BY expressions and tableName available in QueryContext.
- Done.